### ISSUE TITLE
Typecheck fails using go 1.16 modules

### ISSUE THREAD

**Issue by jvrahav** [golangci/golangci-lint#1921]
- [x] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [x] Yes, I've searched similar issues on GitHub and didn't find any.
- [x] Yes, I've included all information below (version, config, etc).
- [x] Yes, I've tried with the standalone linter if available. (https://golangci-lint.run/usage/linters/)

<details><summary>Description of the problem</summary>

typecheck reports erroneous report. For instance, it reports that an import is declared and unused however the import is used
in the code to access a file from the imported package

Similarly there are errors report about unused variables

</details>

<details><summary>Version of golangci-lint</summary>

```
golangci-lint has version v1.39.0 built from 9aea4aee1c0d47f74c016c1f0066cb90e2f7d2e8
```

</details>

<details><summary>Config file</summary>

```console
$ cat .golangci.yml
run:
  # Skip checking the _test files. Enables in phases.
  tests: false

  deadline: 10m
  skip-dirs:
    - vendor
    - pkg
  skip-files:
    # Please keep in alphabetical order.
    - file1
    - file2
linters-settings:
  gocyclo:
    min-complexity: 15
  misspell:
    locale: US
  dupl:
    threshold: 100

linters:
  disable-all: true
  enable:
    - deadcode
    - dupl
    - goconst
    - gocyclo
    - gofmt
    - goimports
    - golint
    - govet
    - ineffassign
    - misspell
    - nakedret
    - structcheck
    - unparam
    - unused
    - varcheck
```
</details>

<details><summary>Go environment</summary>

```console
$ go version && go env
go version go1.16.2b7 linux/amd64
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="usr/go-build"
GOENV="/usr/username/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/usr/username/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/usr/bin/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="linux64"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="linux64/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.16.2b7"
GCCGO="gccgo"
AR="ar"
CC="x86_64-vmk-linux-gnu-gcc"
CXX="/build/mts/linux-centos72-gc32/usr/bin/x86_64-vmk-linux-gnu-g++"
CGO_ENABLED="1"
GOMOD="/dev/null"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/tmp/build2821275998=/tmp/build -gno-record-gcc-switches"
```

</details>

<details><summary>Verbose output of running</summary>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
INFO [config_reader] Used config file .golangci.yml
INFO [lintersdb] Active 14 linters: [deadcode dupl goconst gocyclo gofmt goimports golint govet ineffassign misspell nakedret structcheck unparam varcheck]
INFO [loader] Using build tags: [unit_tests]
INFO [loader] Go packages loading at mode 575 (compiled_files|deps|exports_file|name|types_sizes|files|imports) took 977.874648ms
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 78.415282ms
INFO [linters context/goanalysis] analyzers took 18.54364526s with top 10 stages: buildssa: 5.895955389s, gofmt: 2.936883828s, goimports: 2.448702903s, dupl: 2.017199862s, misspell: 942.555168ms, unparam: 860.516148ms, golint: 848.41798ms, ctrlflow: 228.306019ms, printf: 216.667121ms, ineffassign: 161.256734ms
WARN [runner/source_code] Failed to get line 155 for file $GOROOT/src/io/fs/fs.go: failed to get file $GOROOT/src/io/fs/fs.go lines cache: can't get file $GOROOT/src/io/fs/fs.go bytes from cache: can't read file $GOROOT/src/io/fs/fs.go: open $GOROOT/src/io/fs/fs.go: no such file or directory
WARN [runner/source_code] Failed to get line 151 for file $GOROOT/src/io/fs/fs.go: failed to get file $GOROOT/src/io/fs/fs.go lines cache: can't get file $GOROOT/src/io/fs/fs.go bytes from cache: can't read file $GOROOT/src/io/fs/fs.go: open $GOROOT/src/io/fs/fs.go: no such file or directory
INFO [runner/skip dirs] Skipped 17 issues from dir pkg/util by pattern pkg
INFO [runner/skip dirs] Skipped 28 issues from dir pkg/testutil/govmomi/simulatorxx by pattern pkg
INFO [runner/skip dirs] Skipped 2 issues from dir pkg/vcrestlib by pattern pkg
INFO [runner/skip dirs] Skipped 1 issues from dir pkg/retry by pattern pkg
INFO [runner/skip dirs] Skipped 8 issues from dir pkg/sizing by pattern pkg
INFO [runner/skip dirs] Skipped 7 issues from dir pkg/kubectl/drain by pattern pkg
INFO [runner/skip dirs] Skipped 2 issues from dir pkg/vchelper/types/fake by pattern pkg
INFO [runner/skip dirs] Skipped 2 issues from dir pkg/informer by pattern pkg
INFO [runner/skip dirs] Skipped 1 issues from dir pkg/vchelper/types by pattern pkg
INFO [runner/skip dirs] Skipped 3 issues from dir pkg/logger by pattern pkg
INFO [runner/skip dirs] Skipped 1 issues from dir pkg/testutil by pattern pkg
INFO [runner/skip dirs] Skipped 2 issues from dir pkg/http/handlers by pattern pkg
INFO [runner/skip dirs] Skipped 1 issues from dir pkg/logger/hooks by pattern pkg
INFO [runner] Issues before processing: 10831, after processing: 23
INFO [runner] Processors filtering stat (out/in): path_prettifier: 10831/10831, autogenerated_exclude: 442/10264, exclude-rules: 415/442, max_per_file_from_linter: 23/23, max_from_linter: 23/23, exclude: 442/442, nolint: 384/415, diff: 23/23, max_same_issues: 23/23, filename_unadjuster: 10831/10831, source_code: 23/23, path_shortener: 23/23, sort_results: 23/23, cgo: 10831/10831, skip_files: 10339/10831, skip_dirs: 10264/10339, identifier_marker: 442/442, uniq_by_line: 23/384, severity-rules: 23/23, path_prefixer: 23/23
INFO [runner] processing took 98.395652ms with stages: skip_files: 29.19748ms, nolint: 25.945891ms, path_prettifier: 14.837081ms, exclude-rules: 9.809671ms, identifier_marker: 7.698733ms, autogenerated_exclude: 5.835944ms, skip_dirs: 2.644861ms, cgo: 1.02025ms, source_code: 688.338µs, filename_unadjuster: 655.088µs, uniq_by_line: 37.359µs, max_same_issues: 10.005µs, path_shortener: 7.483µs, max_from_linter: 3.468µs, max_per_file_from_linter: 2.477µs, diff: 475ns, severity-rules: 311ns, exclude: 258ns, sort_results: 254ns, path_prefixer: 225ns
INFO [runner] linters took 1.718334299s with stages: goanalysis_metalinter: 1.619858594s
vclib/guestop.go:12:2: "net/url" imported but not used (typecheck)
	"net/url"
	^
compatibility/check_cluster_compatibility.go:17:2: "github.com/vmware/govmomi/find" imported but not used (typecheck)
	"github.com/vmware/govmomi/find"
	^
kube_supervisor_service_version.go:529:2: log declared but not used (typecheck)
	log := logger.CtxLog(ctx)
	^
../../../../../build/linux64/src/archive/zip/reader.go:639:2: IsDir redeclared (typecheck)
	fs.DirEntry
	^
$GOROOT/src/io/fs/fs.go:155:1: other declaration of IsDir (typecheck)
$GOROOT/src/io/fs/fs.go:151:1: other declaration of Name (typecheck)
INFO File cache stats: 300 entries of total size 2.5MiB
INFO Memory: 29 samples, avg is 411.1MB, max is 753.3MB
INFO Execution took 2.784872089s
```

</details>

<details><summary>code</summary>

```go

import (
"net/url"
)

func name () {
if urlErr, ok := cmdRes.Err.(*url.Error); ok { <--- its being used here
			if urlErr.Err == context.DeadlineExceeded {
				return "", cmdRes.Stderr, locerrors.NewTimeoutError(ctx, locerrors.VCenterWcpClusterGuestOpTimeout,
					personaStr, vmMoref.Value)
			}
		}
}


```

</details>

**Comment by ldez** [comment]
Hello,

do you have `go.mod` file?

could you provide a minimal reproducible example (`go.mod`, `sample.go`)?

**Comment by yaroslavcev** [comment]
👋 I've experienced a similar problem, `go mod tidy` did the trick in my case

**Comment by jvrahav** [comment]
Thanks for the suggestion. i'm building golangci-lint from source locally and using it. 
I realised that i was building golangci-lint with go 1.13.
I built it with go1.14 and everything is working as expected.

However, i have one question out of curiosity,Forgive my ignorance here. can someone help me understand why there is a difference in rule evaluation when i build golangci-lint with go 1.13 as opposed to go 1.14.

**Comment by jvrahav** [comment]
> 👋 I've experienced a similar problem, `go mod tidy` did the trick in my case

Thanks will try that out

**Comment by yalegko** [comment]
It looks like a common error when project (note, not single package) can't be compiled.

Had pretty the same symptoms when I forgot to install necessary cross-compilers for CGO. 

That was a bit tricky to understand, cos errors weren't related to any of CGO packages and were blaming imports, instead the actual errors.

Seems as `typecheck` issue, doesn't it?

**Comment by ldez** [comment]
`typecheck` can't have issues because it's not a real linter: it is just a way to display error from Go compilation.
